### ISSUE TITLE
Create imgur transport using LSB encoding

### ISSUE THREAD

**Issue by Und3rf10w** [Und3rf10w/external_c2_framework#11]
Create a new transport and encoder that utilizes imgur for c2, with images that use [Least significant bit stenography](https://en.wikipedia.org/wiki/Least_significant_bit)

Logic could be something similar to this:

```
                  Sending data
            +------------------------+
            |                        |
            |  Get data from server  |
            |                        |                                                                              +-------------------+
            +---------+--------------+                                              +---------------------+         |                   |
                      |                                                             |                     |         | Transport uploads |
                      |                                                             | Return list of      |         | list of images to |
                      |                                                             | images to transport +-------->+ new Imgur album   |
                      |                                                             |                     |         |                   |          +-----------+
                      |                                                             +----------+----------+         +-------------------+          |           |
                      v                        +---------------------------+                   ^                                                   | Add image +<-------------------+
            +---------+--------------+         |                           |                   |                                                   | to list   |                    |
            |                        |         |                           |                   |                                                   |           |                    |
            | Send data to transport +-------->+ Transport sends raw data  |                   |                                                   +----+------+                    |
            |                        |         |   to encoder for encoding |                   |                                                        |              +------------+-----------------------+
            |                        |         |                           |         +---------+-------------+        +------------------------+        V              |                                    |
            +------------------------+         |                           |         |                       |        |                        |   +-----------+       | Change upperleftmost               |
                                               +------------+--------------+         | Loop through chunks,  |        | Start a counter,       |   |           |       | pixel to:                          |
                                                            |                        | Creating an image for +--------+ Ends when hits:        |   | increment |       | ($VERSION, $COUNTER, $TOTAL_PAGES) |
                                                            |                        | each chunk            |        | (number_of_chunks - 1) +<--+ counter   |       |                                    |
                                                            |                        |                       |        |                        |   |           |       +--------------+---------------------+
                                                            |                        +---------+-------------+        +---------+--------------+   +-----------+                      ^
                                                            |                                  ^                                |                                                     |
                                                            v                                  |                                |                                                     |
                                                +-----------+--------------+        +----------+------------------+             v                 +-------------------+     +---------+---------+
                                                |                          |        |                             |     +-------+-------------+   |                   |     |                   |
                                                | Encoder base64s the data +------->+ Encoder splits data         |     |                     |   | Loop through # of |     | Change Blue value |
                                                |                          |        | into 1079 byte sized chunks |     | Create a new image  +-->+ horizontal pixels +---->+ to chr(byte)      |
                                                +--------------------------+        |                             |     |                     |   | Ignoring first.   |     |                   |
                                                                                    +-----------------------------+     +---------------------+   |                   |     +-------------------+
                                                                                                                                                  +-------------------+




   Retrieving Data
+------------------+
|                  |
| Check Imgur user |
| for new album    |
|                  |
+-------+----------+
        |
        |
        v                  +------------------------+                                     +-------------------+         +-----------------------+
 +------+--------+         |                        |                                     |                   |         |                       |
 |               |         | Download all photos in |                                     | Increment counter |         | Move image to correct |
 | See new album +-------->+ album as a list        |                                     |                   +<--------+ location in new list  |
 |               |         |                        |                                     +--------+----------+         |                       |
 +---------------+         +----------+-------------+                                              |                    +-----------+-----------+
                                      |                                                            |                                ^
                                      |                                                            v                                |
                                      v                                               +------------+---------------+    +-----------+------------+
                          +-----------+-------------+      +--------------------+     |                            |    |                        |
                          |                         |      |                    |     | Start a counter, ends when |    | Read topleftmost       |
                          | Send list of downloaded |      | Reorder the photos +-----+ hits (number_of_photos)    +--->+ pixel of current image |
                          | photos to encoder for   +----->+ in a loop          |     |                            |    |                        |
                          | decoding                |      |                    |     +----------------------------+    +------------------------+
                          |                         |      +-------+------------+
                          +-------------------------+              |
                                                                   |
+----------------+                                                 |
|                |       +---------------------+                   |
| Return the raw |       |                     |                   |
| data to the    |       | Return the raw data |                   |
| application    +<------+ to transport        |                   |
|                |       |                     |                   |
+----------------+       +------------+--------+                   |
                                      ^                            v
                                      |                    +-------+-----------+         +----------------------------+
                                      |                    |                   |         |                            |       +-----------------+
                               +------+------------+       | Loop through new, |         | Start a counter, ends when |       |                 |
                               |                   |       | ordered list of   +---------+ hits (number_of_photos)    +<------+ Increment image |
                               | unbase64 the list +<------+ images            |         |                            |       | counter         |
                               |                   |       |                   |         +-------------+--------------+       |                 |
                               +-------------------+       +-------------------+                       |                      +---------+-------+
                                                                                                       |                                ^
                                                                                                       v                                |
                                                                                           +-----------+------------+          +--------+----------+     +-------------------+
                                                                                           |                        |          |                   |     |                   |
                                                                                           | Start a counter,       |          | Loop through # of |     | append            |
                                                                                           | Ends when hits:        +----------+ horizontal pixels +-----+ ord(pix[x][x][2]) |
                                                                                           | (number_of_chunks - 1) |          | Ignoring first.   |     | to byte list      |
                                                                                           |                        |          |                   |     |                   |
                                                                                           +------------------------+          +------------------++     +---------+---------+
                                                                                                                                                  ^                |
                                                                                                                                                  |                v
                                                                                                                                                  |    +-----------+-------------+
                                                                                                                                                  |    |                         |
                                                                                                                                                  +----+ Increment pixel counter |
                                                                                                                                                       |                         |
                                                                                                                                                       +-------------------------+


```

**Comment by Und3rf10w** [comment]
Created new branch: [imgurC2](https://github.com/Und3rf10w/external_c2_framework/tree/imgurC2) to track development of this feature

**Comment by Und3rf10w** [comment]
Commit 29ab652 add the LSB encoder in `encoder_lsbjpg.py`

**Comment by Und3rf10w** [comment]
Still need to add logic to order the list of images when decoding:

- [ ] Add logic to `encoder_lsbjpg` to order a list of received image when decoding,

**Comment by Und3rf10w** [comment]
Also, current header logic only supports up to 255 images

- [ ] Fix image header numbering to support more images

**Comment by Und3rf10w** [comment]
Commit 20e8eae add logic to `encoder.decode()` that orders the images before decoding

**Comment by Und3rf10w** [comment]
Imgur's authorization logic is a bit weird because for authentication actions, it will require that the user clicks a link to authenticate and receive an access and refresh tokens.

Obviously, this can't be done on a client we can't communicate with.

We can use our header as a way to store this info though.

if `pix[0,0][0] == 1`, then we know that this image contains the tokens we need. The client will be able to view this image containing the tokens without authenticating, while maintaining our opsec.

Once the client has received the tokens, it can then trigger an acknowledgement to the server by deleting the album with the tokens.

Only caveat is that the operator will have to configure the application to accept when starting the server.

All this logic could look something like this:

```
                                  +----------------------------+
                                  |                            |        +-----------------------------+
  Server's prepTransport()        |  Add the following header: |        |                             |
                                  |  pix[0,0] = (1, 1, 1)      +------->+  Insert the list of tokens  |
+-----------------------------+   |                            |        |  across the diagonal pixels |
|                             |   +---------+------------------+        |  ignoring first             |
| Server requests that        |             |                           |                             |
| user authorizes application |             |                           +-----------------------------+
|                             |             |
+--------+--------------------+             |
         |                                  |                                  Client's prepTransport()
         v                       +----------+----------+
+--------+---------------+       |                     |                 +------------------------------+
|                        |       | Server sends        |                 |                              |
| User inputs auth token +------>+ tokens as new album |                 | Client sees new single-photo +<-----------+
|                        |       |                     |                 | album                        |            |
+------------------------+       +----------+----------+                 |                              |            |
                                            |                            +------------------------------+            |
                +--------------------+      |                                        |                               |
                |                    |      |                                        |                               |
                |                    v      v                                        |                               |
          +-----+-------+       +----+------+----------+                             v                               |
          |             |       |                      |                 +-----------+-------------+       +---------+---------+
          | Album still |       | Server checks to see |                 |                         |       |                   |
          | exists      +<------+ if album was deleted |                 | Client checks value of  +------>+  pix[0,0][0] != 1 |
          |             |       |                      |                 | pix[0,0][0] == 1        |       |                   |
          +-------------+       +----------+-----------+                 |                         |       +-------------------+
                                           |                             +----------+--------------+
                                           |                                        |
                                           |                                        |
                                           v                                        v                                               +---------------+
        +---------------+        +---------+--------+                      +--------+---------+             +--------------+        |               |
        |               |        |                  |                      |                  |             |              |        | Continue main |
        | Continue main |        |  Album no longer |                      | pix[0,0][0] == 1 |             | Delete album +------->+ application   |
        | application   +<-------+  exists          |                      |                  |             |              |        | logic         |
        | logic         |        |                  |                      +--------+---------+             +------+-------+        |               |
        |               |        +------------------+                               |                              ^                +---------------+
        +---------------+                                                           v                              |
                                                                          +---------+--------------+               |
                                                                          |                        |         +-----+----------------------+
                                                                          | Iterate through pixels |         |                            |
                                                                          | diagonally, ignoring   +-------->+ Extract access and refresh |
                                                                          | first                  |         | tokens from photo          |
                                                                          |                        |         |                            |
                                                                          +------------------------+         +----------------------------+

```

**Comment by Und3rf10w** [comment]
- [ ] Create logic to handle transmission of access and refresh tokens between client and server

**Comment by Und3rf10w** [comment]
Moving tracking of development for this to #12

**Comment by Und3rf10w** [comment]
Fixed by #12
